File: src/net/dinkla/raytracer/btdf/PerfectTransmitter.java
package net.dinkla.raytracer.btdf;

import net.dinkla.raytracer.colors.Color;
import net.dinkla.raytracer.hits.Shade;
import net.dinkla.raytracer.math.Normal;
import net.dinkla.raytracer.math.Vector3D;

/**
 * Created by IntelliJ IDEA.
 * User: Jörn Dinkla
 * Date: 01.05.2010
 * Time: 09:56:21
 * To change this template use File | Settings | File Templates.
 */
public class PerfectTransmitter<C extends Color> extends BTDF<C> {

    public double ior;
    public double kt;

    public PerfectTransmitter() {
        kt = 1.0;
        ior = 1.0;
    }
    
    @Override
    public C f(Shade sr, Vector3D wo, Vector3D wi) {
        throw new RuntimeException("PerfectTransmitter.f");
    }

    @Override
    public C rho(Shade sr, Vector3D wo) {
        throw new RuntimeException("PerfectTransmitter.rho");
    }

    @Override
    public Sample sampleF(Shade sr, Vector3D wo) {
        Sample result = new Sample();
        Normal n = sr.getNormal();
        double cosThetaI = n.dot(wo);
        double eta = ior;
        if (cosThetaI < 0) {
            cosThetaI = -cosThetaI;
            n = n.negate();
            eta = 1.0 / eta;
        }
        double cosThetaTSqr = 1.0 - (1.0 - cosThetaI * cosThetaI) / (eta * eta);
        double cosThetaT = (double) Math.sqrt(cosThetaTSqr);
        result.wt = wo.mult(-eta).minus(n.mult(cosThetaT - cosThetaI / eta));
        double f1 = kt / (eta*eta);
        double f2 = sr.getNormal().dot(result.wt);
        result.color = (C) C.WHITE.mult(f1 / Math.abs(f2));
        return result;
    }

    @Override
    public boolean isTir(Shade sr) {
        Vector3D wo = sr.ray.getD().mult(-1);
        double cosThetaI = wo.dot(sr.getNormal());
        double eta = ior;
        if (cosThetaI < 0) {
            eta = 1.0 / eta;
        }
        double cosThetaTSqr = 1.0 - (1.0 - cosThetaI * cosThetaI) / (eta * eta);
        return cosThetaTSqr < 0;
    }
    
}

File: src/net/dinkla/raytracer/btdf/PerfectTransmitter.kt
package net.dinkla.raytracer.btdf

import net.dinkla.raytracer.colors.Color
import net.dinkla.raytracer.hits.Shade
import net.dinkla.raytracer.math.Vector3D

class PerfectTransmitter : BTDF() {

    var ior: Double = 0.toDouble()
    var kt: Double = 0.toDouble()

    init {
        kt = 1.0
        ior = 1.0
    }

    override fun f(sr: Shade, wo: Vector3D, wi: Vector3D): Color {
        throw RuntimeException("PerfectTransmitter.f")
    }

    override fun rho(sr: Shade, wo: Vector3D): Color {
        throw RuntimeException("PerfectTransmitter.rho")
    }

    override fun sampleF(sr: Shade, wo: Vector3D): BTDF.Sample {
        val result = newSample()
        var n = sr.normal
        var cosThetaI = n.dot(wo)
        var eta = ior
        if (cosThetaI < 0) {
            cosThetaI = -cosThetaI
            n = n.negate()
            eta = 1.0 / eta
        }
        val cosThetaTSqr = 1.0 - (1.0 - cosThetaI * cosThetaI) / (eta * eta)
        val cosThetaT = Math.sqrt(cosThetaTSqr)
        result.wt = wo.times(-eta).minus(n.mult(cosThetaT - cosThetaI / eta))
        val f1 = kt / (eta * eta)
        val f2 = sr.normal.dot(result.wt!!)
        result.color = Color.WHITE.times(f1 / Math.abs(f2))
        return result
    }

    override fun isTir(sr: Shade): Boolean {
        val wo = sr.ray.d.times(-1.0)
        val cosThetaI = wo.dot(sr.normal)
        var eta = ior
        if (cosThetaI < 0) {
            eta = 1.0 / eta
        }
        val cosThetaTSqr = 1.0 - (1.0 - cosThetaI * cosThetaI) / (eta * eta)
        return cosThetaTSqr < 0
    }

}
